overlord/hookstate/hookmgr.go

```diff
@@ -143,13 +153,38 @@ func (m *HookManager) doRunHook(task *state.Task, tomb *tomb.Tomb) error {
 		return err
 	}
 
-	// TODO: Actually dispatch the hook.
+	_, err = RunHook(setup.Snap, setup.Revision, setup.Hook)
+	if err != nil {
+		exitErr, ok := err.(*exec.ExitError)
+		if ok {
+			// "exit code 1" is not particularly helpful. So rewrite the error
+			// using stderr instead.
+			err = fmt.Errorf("%s", strings.Trim(string(exitErr.Stderr), " \n"))
+		}
+
+		if handlerErr := handler.Error(err); handlerErr != nil {
+			return handlerErr
+		}
+
+		return err
+	}
 
-	// Done with the hook. TODO: Check the result, if success call Done(), if
-	// error, call Error(). Since we have no hooks, for now we just call Done().
-	if err := handler.Done(); err != nil {
+	if err = handler.Done(); err != nil {
 		return err
 	}
 
 	return nil
 }
+
+func doRunHook(snapName string, revision snap.Revision, hookName string) ([]byte, error) {
+	revisionString := ""
+	if !revision.Unset() {
```

Comment:
We _can_ ensure that, but currently if `snap run` doesn't receive a revision it simply uses the active one, which I think is a sensible default. As a result, the code behind `snap set` won't need to look up the active revision, it just doesn't pass one. Do you want the revision lookup code in both places?